doc/automaatiotestit.md

```diff
@@ -0,0 +1,215 @@
+
+Lähtökohtaisesti testejä on pyritty rajaamaan suhteellisen pieniksi, sillä testattavan tilanteen tulkinta on
+yksinkertaisempaa, kun testattava ominaisuusmäärä on rajatumpi. Vastaavasti jos koodikantaan tehdyt muutokset
+aiheuttavat testin epäonnistumisen muutosten jälkeen, on usein yksiselitteisempää huomata ongelma jos vain yksi testi
+epäonnistuu verrattuna vaikkapa moneen kymmeneen.
+
+Testien tulisi alustaa kaikki tarvitsemansa data tai tila testin alussa, jotta ylimääräisiä riippuvuuksia testien
+välille ei synny ja testattavan tilan tulkinta on helpompaa. Geoviitteen koodikanta kuitenkin sisältää eritasoisia
+testidatan luontiin tarkoitettuja funktioita, jotka helpottavat testikoodin kirjoittajaa muodostamaan haluamansa
+tilanteen sekä usein auttavat myös testitilanteen tulkinnassa. Tavoitteena olisi, että mitä yksinkertaisempi tilanne,
+sitä yksinkertaisempia datanalustusfunktioita hyödynnetään. Jos taas testattava tilanne on monimutkaisempi, käytetään
+laajemman testattavan tilan muodostavia funktioita. Näitä voi myös muodostaa lisää tarpeiden mukaisesti.
+
+Testidatan alustuksen tavoitteena olisi välttää käyttämästä liian suuren tilanteen alustusta yksinkertaisen tilanteen
+testaamista varten: testikoodin pitäisi alustaa minimaalinen määrä dataa, jotta testi on mahdollista suorittaa.
+Minimaalisen tilan muodostamisella pyritään välttämään riippuvuussuhteiden muodostumista testien välillä. Jos
+esimerkiksi parikymmentä testiä hyödyntäisi samaa testidataa, ja yksi testi tarvitsisikin hieman laajemman tilan, niin
+tämän pienen tilamuutoksen tekeminen testidataan aiheuttaisi myös lisätilan päätymisen kaikkiin muihinkin testeihin,
+jotka hyödyntäisivät samaa testidataa. Tämä taas saattaa aiheuttaa testibugeja tai muutosedellytyksiä muihinkin
+testeihin, jotka hyödyntäisivät tätä testidatajoukkoa. Esimerkiksi jokin toinen samaa testidataa hyödyntävä testi ei
+välttämättä menisi enää ilman muutoksia läpi, tai se testaisi väärää tilannetta eli se aiheuttaisi esimerkiksi vääriä
+positiivisia tai vääriä negatiivisia testituloksia (tyypin I ja II virheitä, hylkäämis- ja hyväksymisvirheitä).
+
+Rakenteellisena tavoitteena yksinkertaisia (yksikkötestejä) olisi määrällisesti eniten, hieman monimutkaisempia testejä
+(integraatiotestejä) vähemmän, ja monimutkaisia, laajoja testejä (E2E-testejä) olisi vähiten. On usein helpompaa tulkita
+yksittäisen yksikkötestien epäonnistumisesta mikä meni vikaan, kuin tulkita E2E-testin epäonnistumisesta miksi jokin
+tietty pieni asia sattuikin epäonnistumaan tietyn E2E-testin suorituksen aikana.
+
+## Testikoodista
+
+Kotlin-koodin testit hyödyntävät Jupiter-kirjastoa.
+
+Testiluokat sisältävät annotointeja riippuen hieman minkätyyppistä ympäristöä ne tarvitsevat. Jokainen testiluokka
+kuitenkin annotoidaan `@SpringBootTest`:llä niiden tunnistamista varten, sekä usein testiluokille määritellään myös
+Springin ympäristöön liittyviä konfiguraatioannotointeja, kuten esimerkiksi `@ActiveProfiles("dev", "test")`.
```

Comment:
Tämä annotaatiojuttu (ainakin sikäli kun olen ymmärtänyt oikein) pätee ainoastaan IT-testeihin, yksikkötesteissä ei käytetä `@SpringBootTest`- tai mitään muutakaan annotaatiota